bin/node-template/node/src/service.rs

```diff
@@ -99,19 +85,35 @@ pub fn new_full(config: Configuration) -> Result<impl AbstractService, ServiceEr
 	let name = config.network.node_name.clone();
 	let disable_grandpa = config.disable_grandpa;
 
-	let (builder, mut import_setup, inherent_data_providers) = new_full_start!(config);
+	let (
+		client, mut import_setup, inherent_data_providers, backend, task_manager, keystore,
+		import_queue, select_chain, transaction_pool, background_tasks,
+	) = new_full_start!(config);
 
 	let (block_import, grandpa_link) =
 		import_setup.take()
 			.expect("Link Half and Block Import are present for Full Services or setup failed before. qed");
 
-	let service = builder
-		.with_finality_proof_provider(|client, backend| {
-			// GenesisAuthoritySetProvider is implemented for StorageAndProofProvider
-			let provider = client as Arc<dyn StorageAndProofProvider<_, _>>;
-			Ok(Arc::new(GrandpaFinalityProofProvider::new(backend, provider)) as _)
-		})?
-		.build()?;
+	let provider = client.clone() as Arc<dyn StorageAndProofProvider<_, _>>;
+	let finality_proof_provider = Arc::new(GrandpaFinalityProofProvider::new(backend.clone(), provider));
+	let service = sc_service::build(sc_service::ServiceDescriptor {
```

Comment:
Essentially we have replaced the builder with a almost 20 lines initialization of a struct. Where is the improvement?

So I know that the current builder is shit, but if we do it now, we should do it correct and not introduce another "intermediate state".

Reply:
This resolves a lot of the issues @gnunicorn talks about here: https://github.com/paritytech/substrate/issues/4587#issue-547973409.